pallets/amm/src/lib.rs

```diff
@@ -465,3 +483,127 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // subtract lp fees from amount_in minus protocol fees
+                let amount_in_after_all_fees = amount_without_protocol_fees
+                    .checked_sub(lp_fees)
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // 3. Given the input amount amount_in left after fees, compute amount_out
+                // let amount_out = amount_in * supply_out / (supply_in + amount_in)
+                let amount_out = amount_in_after_all_fees
+                    .saturating_mul(supply_out)
+                    .checked_div(
+                        supply_in
+                            .checked_add(amount_in_after_all_fees)
+                            .ok_or(ArithmeticError::Overflow)?,
+                    )
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // 4. If `amount_out` is lower than `min_amount_out`, error
+                ensure!(
+                    amount_out >= minimum_amount_out && amount_in > 0,
+                    Error::<T, I>::InsufficientAmountOut
+                );
+
+                // 5. Update the `Pools` storage to track the `base_amount` and `quote_amount`
+                // variables (increase and decrease by `amount_in` and `amount_out`)
+                // increase liquidity_amount.base_amount by amount_in, unless inverted
+                if is_inverted {
+                    liquidity_amount.quote_amount = liquidity_amount
+                        .quote_amount
+                        .checked_add(amount_without_protocol_fees)
+                        .ok_or(ArithmeticError::Overflow)?;
+
+                    liquidity_amount.base_amount = liquidity_amount
+                        .base_amount
+                        .checked_sub(amount_out)
+                        .ok_or(ArithmeticError::Underflow)?;
+                } else {
+                    liquidity_amount.base_amount = liquidity_amount
+                        .base_amount
+                        .checked_add(amount_without_protocol_fees)
+                        .ok_or(ArithmeticError::Overflow)?;
+
+                    liquidity_amount.quote_amount = liquidity_amount
+                        .quote_amount
+                        .checked_sub(amount_out)
+                        .ok_or(ArithmeticError::Underflow)?;
+                }
+                *pool_liquidity_amount = Some(liquidity_amount);
+
+                // 6. Wire amount_in of the input token (identified by pair.0) from who to PalletId
+                T::Currency::transfer(
+                    pair.0,
+                    &who,
+                    &Self::account_id(),
+                    amount_without_protocol_fees,
+                )?;
+
+                // 7. Wire amount_out of the output token (identified by pair.1) to who from PalletId
+                T::Currency::transfer(pair.1, &Self::account_id(), &who, amount_out)?;
```

Comment:
```suggestion
                T::Currency::transfer(quote_asset, &Self::account_id(), &who, amount_out)?;
```

Reply:
I don't think that `pair.0` and `quote_asset` are always the same. 

We know that `pair.0` will be the input currency from the trade, where `quote_asset` is the RHS asset in our pool. It's possible that a trader will want to input the `quote_asset` of a pool and get the `base_asset` in return. 

Due to this we can not assume that `pair.0 === quote_asset`

Reply:
Can we extract variables from pair? using pair.1 and pair.2 doesn't look good. I also used in the past but team didn't like this. Please see this https://github.com/parallel-finance/parallel/pull/450#discussion_r695339486

Reply:
yep thats a good idea 👌